core/error.go

```diff
@@ -0,0 +1,29 @@
+package core
+
+import "github.com/vektah/gqlparser/v2/ast"
+
+type Error struct {
+	Query *Query `json:"-"`
+
+	Message string `field:"true" doc:"A description of the error."`
+
+	// Stack    string `field:"true" doc:"The stack trace of the error."`
+	// Source   *File  `field:"true" doc:"The source file where the error occurred."`
+	// Line     int    `field:"true" doc:"The line in the source file where the error occurred."`
+	// Original *Error `field:"true" doc:"The original error that caused this error."`
+
+	// what about screenshots and stuff?
```

Comment:
Hmm we could actually do this in a somewhat strongly typed way using IDs, since they embed their expected type. You could technically take any ID, context-free, parse it to discover the type, and then call `loadFooTypeFromID(id) { ... }` to interact with the object. 

But, I guess that only helps with storing objects, as opposed to scalars.

Really depends on what we need from this IMO in terms of real world use cases - I'm hesitant to jump straight to a generic map for example, since it's hard to do something high-level + useful for that (e.g. display screenshots for browser test failures in Cloud) without defining some sort of schema/semantic conventions. Might be better to just have hyper specific things like `ImagesDir *Directory`. Super low conviction though.

Tangent: in "pure GraphQL" the closest thing is just having a generic `ID` scalar type and doing something like this (see [Global Object Identification](https://graphql.org/learn/global-object-identification/)):

```graphql
{
  node(id: "DSFKjhsckdjhsdkjv") {
    ... on Directory {
      export(path: ".")
    }
  }
}
```

But `Node` interface is `id: ID!`, and we prefer `id: DirectoryID!` for codegen purposes, so we have all these `loadFooFromID` APIs instead.

Reply:
> Really depends on what we need from this IMO in terms of real world use cases

Yeah, agreed, happy to defer on this!

> Might be better to just have hyper specific things like `ImagesDir *Directory`. Super low conviction though.

Ooh, I do not like the idea of hyper-specific stuff in the API like that. Not quite sure what a alternative is though, hm, without punting all the hard problems into "cloud".

If we could though (:laughing:), maybe cloud could have a series of "matchers" to decide how to display generic dagger api objects?

- a `Container` gets a list of layers
- a `File` with a `.png`/`.jpeg`/etc filename extension gets displayed in an `<img>` tag
- a `File` with a `.md` filename is rendered as a markdown file
- a `Directory` is just displayed as the sum of it's parts
- a future magical `Report` type gets rendered as a table

This would be *pretty* neat, and could also help solve how any return type could get rendered in cloud (which I feel like we're gonna need going into pipeline results around future planned ci stuff?).

Just brainstorming really, not really related to this issue anymore.